packages/core/admin/ee/server/services/review-workflows/stages.js

```diff
@@ -51,63 +51,52 @@ module.exports = ({ strapi }) => {
       return strapi.entityService.count(STAGE_MODEL_UID);
     },
 
-    async replaceWorkflowStages(workflowId, stages) {
+    async replaceWorkflowStages(workflowId, newStages) {
       const workflow = await workflowsService.findById(workflowId, { populate: ['stages'] });
 
-      const { created, updated, deleted } = getDiffBetweenStages(workflow.stages, stages);
+      const { created, updated, deleted } = getDiffBetweenStages(workflow.stages, newStages);
 
       assertAtLeastOneStageRemain(workflow.stages, { created, deleted });
 
       return strapi.db.transaction(async () => {
-        const defaultWorkflow = await getDefaultWorkflow({ strapi });
-        const newStages = await this.createMany(created, { fields: ['id'] });
-        const stagesIds = stages.map((stage) => stage.id ?? [...newStages].shift().id);
-
+        // Create the new stages
+        const createdStages = await this.createMany(created, { fields: ['id'] });
+        // Put all the newly created stages ids
+        const createdStagesCopy = [...createdStages];
+        const stages = newStages.map((stage) => (stage.id ? stage : createdStagesCopy.shift()));
+        const contentTypes = getContentTypeUIDsWithActivatedReviewWorkflows(strapi.contentTypes);
+
+        // Update the workflow stages
         await mapAsync(updated, (stage) => this.update(stage.id, stage));
 
-        const entitiesToMove = [];
+        // Delete the stages that are not in the new stages list
         await mapAsync(deleted, async (stage) => {
-          // Find any entities related to this stage
-          const stageInfo = await this.findById(stage.id, {
-            populate: ['related'],
-          });
-
-          if (!stageInfo?.related?.length) {
-            // If there are no related entities, just delete the stage
-            return this.delete(stage.id);
-          }
-
-          // This stage has related entities that need to be moved to their
-          // target stage
           // Find the nearest stage in the workflow and newly created stages
           // that is not deleted, prioritizing the previous stages
-          const targetStageId = findNearestMatchingStageID(
-            [...defaultWorkflow.stages, ...newStages],
-            defaultWorkflow.stages.findIndex((s) => s.id === stage.id),
+          const nearestStage = findNearestMatchingStage(
+            [...workflow.stages, ...createdStages],
```

Comment:
Maybe we can use the `stages` array here?

Reply:
stages do not contain the original deleted stages, which are needed to find the nearest stage :( Still, I would love to find some workaround to only use stages here.

Reply:
If we introduce reordering in the stages, this will implementation will not work anymore.

Reply:
is the original deleted stages needed to find the nearest stage ? 🤔

Reply:
We need to know the index of the deleted stage. Can't we get that information from elsewhere?

Reply:
yes you can get the index, but you can delete multiple stages at the same time, and that it makes more complicated.

If you have:
stages = [1,2,3,4]

And you remove 2 & 3
stages = [1,4]

The deleted idx of `stage 3` is `idx=1`, so the nearest stage in the new array would be `stage 4`, when it should actually be `stage 1`

Side comment, I would just assign to any entity the first stage of the workflow, or ask the user which stage wants to be replaced.